src/chain.h

```diff
@@ -596,65 +601,100 @@ class CDiskBlockIndex : public CBlockIndex
+    virtual CBlockIndex *Tip() const REQUIRES(cs_main) {
         return vChain.size() > 0 ? vChain[vChain.size() - 1] : NULL;
     }
     
     /** Returns the last tip of the chain, or NULL if none. */
-    CBlockIndex *LastTip() const {
+    virtual CBlockIndex *LastTip() const REQUIRES(cs_main) {
         return vChain.size() > 0 ? lastTip : NULL;
     }
 
     /** Returns the index entry at a particular height in this chain, or NULL if no such height exists. */
-    CBlockIndex *operator[](int nHeight) const {
-        if (nHeight < 0 || nHeight >= (int)vChain.size())
-            return NULL;
-        return vChain[nHeight];
+    virtual CBlockIndex *operator[](int nHeight) const REQUIRES(cs_main) {
+        return at(nHeight);
     }
 
     /** Compare two chains efficiently. */
-    friend bool operator==(const CChain &a, const CChain &b) {
-        return a.vChain.size() == b.vChain.size() &&
-               a.vChain[a.vChain.size() - 1] == b.vChain[b.vChain.size() - 1];
+    friend bool operator==(const CChain &a, const CChain &b) REQUIRES(cs_main) {
+        return a.Height() == b.Height() &&
+               a.LastTip() == b.LastTip();
     }
 
     /** Efficiently check whether a block is present in this chain. */
-    bool Contains(const CBlockIndex *pindex) const {
+    virtual bool Contains(const CBlockIndex *pindex) const REQUIRES(cs_main) {
         return (*this)[pindex->GetHeight()] == pindex;
     }
 
     /** Find the successor of a block in this chain, or NULL if the given index is not found or is the tip. */
-    CBlockIndex *Next(const CBlockIndex *pindex) const {
+    virtual CBlockIndex *Next(const CBlockIndex *pindex) const REQUIRES(cs_main) {
         if (Contains(pindex))
             return (*this)[pindex->GetHeight() + 1];
         else
             return NULL;
     }
 
     /** Return the maximal height in the chain. Is equal to chain.Tip() ? chain.Tip()->GetHeight() : -1. */
-    int Height() const {
+    virtual int Height() const REQUIRES(cs_main) {
         return vChain.size() - 1;
     }
 
     /** Set/initialize a chain with a given tip. */
-    void SetTip(CBlockIndex *pindex);
+    virtual void SetTip(CBlockIndex *pindex) REQUIRES(cs_main);
 
     /** Return a CBlockLocator that refers to a block in this chain (by default the tip). */
-    CBlockLocator GetLocator(const CBlockIndex *pindex = NULL) const;
+    virtual CBlockLocator GetLocator(const CBlockIndex *pindex = NULL) const REQUIRES(cs_main);
 
     /** Find the last common block between this chain and a block index entry. */
-    const CBlockIndex *FindFork(const CBlockIndex *pindex) const;
+    virtual const CBlockIndex *FindFork(const CBlockIndex *pindex) const REQUIRES(cs_main);
+};
+
+#ifdef DEBUG_LOCKORDER
+/************
+ * A mutex-protected version of CChain
+ * Probably good for verifying locking, but no effort
+ * was made to examine efficiency
+ */
+template<class Mutex>
+class MultithreadedCChain : public CChain
```

Comment:
In fact this is not quite a 'multi threaded' chain as it does not provide synchronisation but just checks in the debug mode that access is synced. Just saying